src/consensus/progress_map.zig

```diff
@@ -0,0 +1,1169 @@
+
+        const propagated_stats = try self.propagated_stats.clone(allocator);
+        errdefer propagated_stats.deinit(allocator);
+
+        const replay_stats = try self.replay_stats.arc_ed.rwlock_ed.clone(allocator);
+        errdefer replay_stats.deinit(allocator);
+
+        return .{
+            .is_dead = self.is_dead,
+            .fork_stats = fork_stats,
+            .propagated_stats = propagated_stats,
+            .replay_stats = .{ .arc_ed = .{ .rwlock_ed = replay_stats } },
+            .replay_progress = self.replay_progress,
+            .retransmit_info = self.retransmit_info,
+            .num_blocks_on_fork = self.num_blocks_on_fork,
+            .num_dropped_blocks_on_fork = self.num_dropped_blocks_on_fork,
+        };
+    }
+};
+
+pub const ForkStats = struct {
+    fork_stake: consensus.Stake,
+    total_stake: consensus.Stake,
+    block_height: u64,
+    has_voted: bool,
+    is_recent: bool,
+    is_empty: bool,
+    vote_threshold: VoteThreshold,
+    is_locked_out: bool,
+    voted_stakes: consensus.VotedStakes,
+    duplicate_confirmed_hash: Hash,
+    computed: bool,
+    lockout_intervals: LockoutIntervals,
+    bank_hash: Hash,
+    my_latest_landed_vote: Slot,
+
+    pub const VoteThreshold = std.ArrayListUnmanaged(consensus.ThresholdDecision);
+
+    pub fn deinit(
+        self: ForkStats,
+        allocator: std.mem.Allocator,
+    ) void {
+        var vote_threshold = self.vote_threshold;
+        vote_threshold.deinit(allocator);
+
+        var voted_stakes = self.voted_stakes;
+        voted_stakes.deinit(allocator);
+
+        const lockout_intervals = self.lockout_intervals;
+        lockout_intervals.deinit(allocator);
+    }
+
+    pub fn clone(
+        self: ForkStats,
+        allocator: std.mem.Allocator,
+    ) std.mem.Allocator.Error!ForkStats {
+        var vote_threshold = try self.vote_threshold.clone(allocator);
+        errdefer vote_threshold.deinit(allocator);
+
+        var voted_stakes = try self.voted_stakes.clone(allocator);
+        errdefer voted_stakes.deinit(allocator);
+
+        const lockout_intervals = try self.lockout_intervals.clone(allocator);
+        errdefer lockout_intervals.deinit(allocator);
+
+        return .{
+            .fork_stake = self.fork_stake,
+            .total_stake = self.total_stake,
+            .block_height = self.block_height,
+            .has_voted = self.has_voted,
+            .is_recent = self.is_recent,
+            .is_empty = self.is_empty,
+            .vote_threshold = vote_threshold,
+            .is_locked_out = self.is_locked_out,
+            .voted_stakes = voted_stakes,
+            .duplicate_confirmed_hash = self.duplicate_confirmed_hash,
+            .computed = self.computed,
+            .lockout_intervals = lockout_intervals,
+            .bank_hash = self.bank_hash,
+            .my_latest_landed_vote = self.my_latest_landed_vote,
+        };
+    }
+};
+
+pub const PropagatedStats = struct {
```

Comment:
Could you support `add_vote_pubkey` and `add_node_pubkey`?

Reply:
52b42b82333ca9ebd4cca7bc534cefb46f43e532
de067e322dee7f764a49423e5e9f633057274109